python-cfd/CG.py

```diff
@@ -0,0 +1,66 @@
+import sys
+import math
+import numpy as np
+import scipy as sp
+from scipy import linalg
+
+A = np.array([[-2,1,1,0],[1,-2,0,1],[1,0,-2,1],[0,1,1,-2]])
+b = np.array([1, -1, 1, -1])
+
+# Provide some information
+print("# matrix:")
+for r in A:
+    print("   ", r)
+print("  eigenvalues:", ", ".join(["{:.6g}".format(e) for e in np.linalg.eigvals(A)]))
+print("  rank:", np.linalg.matrix_rank(A))
+
+kerA = sp.linalg.null_space(A)
+print("  nullspace:")
+for r in kerA:
+    print("   ", r)
+
+# Initialize containers and values
+print("# intialization:")
+x = np.array([0., 0., 0., 0.])
+print("  initial guess:", x)
+
+residuals = [b - A @ x]
+rms2 = residuals[0] @ residuals[0]
+print("  residual:{} (L2 norm: {})".format(residuals[0], rms2))
+
+directions = residuals[:]
+print("  directions", [list(d) for d in directions])
+alpha, beta = [], []
+
+# Iterate
+for k in range(min(A.shape)):
+    kp1 = k + 1
+    print("# iteration:", kp1)
+
+    # Compute step
+    alpha.append(rms2 / (directions[k] @ A @ directions[k]))
+    print("  alpha[{}]: {}".format(k, alpha[k]))
+
+    # Update solution
+    x += alpha[k] * directions[k]
+    print("  updated solution:", x)
+
+    # Update residual
+    residuals.append(b - A @ x)
+    new_rms2 = residuals[kp1] @ residuals[kp1]
+    print("  residual:{} (L2 norm: {})".format(residuals[kp1], new_rms2))
+
+    # Terminate early when possible
+    if new_rms2 < 1e-12:
+        print("# CG converged to solution: {} with residual norm: {}".format(
+            x, math.sqrt(new_rms2)))
+        sys.exit(0)
+
+    # Compute new direction
+    beta.append(new_rms2 / rms2)
+    print("  beta[{}]: {}".format(k, beta[k]))
+    directions.append(residuals[kp1] - beta[k] * directions[k])
```

Comment:
why there is a minus sign here? shouldn't this be `residual + ...`? http://www.onmyphd.com/?p=conjugate.gradient.method

Reply:
Because this webpage is incorrect :)

You do need to subtract along the direction -- not add. So either you stick the `-` sign in beta and you can add, or you don't and in this case you need to subtract.

Reply:
https://en.wikipedia.org/wiki/Conjugate_gradient_method

Reply:
![image](https://user-images.githubusercontent.com/18708772/129383564-029e56d4-3c3b-477e-84c3-b7b0cf1b6251.png)

Reply:
I see this, but this is incorrect. This must be a subtraction, not an addition.
When it comes to maths, Wikipedia/random webpages are more often than not, wrong.

Do the math you will see.

Reply:
Check this out, in same wikipedia as the one you sent, a few paragraphs above the pseudo-code:

<img width="256" alt="Screen Shot 2021-08-13 at 5 54 03 PM" src="https://user-images.githubusercontent.com/17439260/129386194-fe8d7fe1-a5fe-405c-8983-b87375c172ae.png">

Reply:
By the way @fnrizzi if you replace the - with a + the algo diverges immediately

Reply:
i agree with you :) I was pointing out how many wrong things there are!

Reply:
You know what? I think I now where this error is coming from.
It's an online book of J. Schwechuck from 1994, that made its way into Wiki....

Reply:
@fnrizzi can you please run it with other examples if you have a minute?

I derived the algorithm by hand myself, there might be errors anyway.

Reply:
Good for me I tested it with the 3 by 3 case and it works.